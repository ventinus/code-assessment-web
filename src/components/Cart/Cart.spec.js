import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { Cart, CartProduct, PrimaryButton } from '..'

const productFactory = (id, title = 'Title', price = 10, quantity = 2, inventory = 4) => ({
  id,
  title,
  price,
  quantity,
  inventory,
})

let actions
const setup = (total, products = []) => {
  actions = {
    onCheckoutClicked: jest.fn(),
  }

  const component = shallow(
    <Cart products={products} total={total} {...actions} />
  )

  return {
    component: component,
    actions: actions,
    button: component.find('.cart__checkout-btn'),
    products: component.find(CartProduct),
    em: component.find('em'),
    total: component.find('.cart__breakdown__value').get(0),
  }
}

describe('Cart component', () => {
  it('renders correctly with no products', () => {
    const tree = renderer
      .create(<Cart products={[]} total={0} {...actions} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should display add some products message', () => {
    const { em } = setup()
    expect(em.text()).toMatch(/^Please add some products to your cart./)
  })

  it('should not render checkout button when no products selected', () => {
    const { component } = setup()
    expect(component.exists('.cart__checkout-btn')).toEqual(false)
  })

  describe('when given product', () => {
    const product = [
      productFactory(1, 'Product 1'),
      productFactory(2, 'Product 2'),
      productFactory(3, 'Product 3'),
    ]

    it('renders correctly with products', () => {
      const tree = renderer
        .create(<Cart products={product} total={76} {...actions} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render products', () => {
      const { products } = setup(9.99, product)
      const props = product[0]

      expect(products).toHaveLength(3)
      expect(products.get(0).props).toEqual(props)
    })

    it('should display total', () => {
      const { total } = setup(76, product)
      expect(total.props.children).toMatch(/^\$76.00/)
    })

    it('should render checkout button', () => {
      const { button } = setup(9.99, product)
      expect(button.exists()).toEqual(true)
    })

    it('should call action on button click', () => {
      const { button, actions } = setup(9.99, product)
      button.simulate('click')
      expect(actions.onCheckoutClicked).toBeCalled()
    })
  })
})
