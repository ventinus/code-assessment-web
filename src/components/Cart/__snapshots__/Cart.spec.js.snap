// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component renders correctly with no products 1`] = `
<div
  className="cart"
>
  <div
    className="border-section header"
    style={
      Object {
        "borderBottomWidth": 1,
      }
    }
  >
    <h1
      className="type--a1"
    >
      Your Cart
    </h1>
  </div>
  <div
    className="cart__no-products"
  >
    <i
      className="icon icon--cart icon--15 icon--empty-cart"
    />
    <em
      className="cart__no-products__copy type--a6 type--block"
    >
      Please add some products to your cart.
    </em>
  </div>
</div>
`;

exports[`Cart component when given product renders correctly with products 1`] = `
<div
  className="cart"
>
  <div
    className="border-section header"
    style={
      Object {
        "borderBottomWidth": 1,
      }
    }
  >
    <h1
      className="type--a1"
    >
      Your Cart
    </h1>
  </div>
  <ul
    className="cart__products"
  >
    <li
      className="cart-product"
    >
      <div
        className="cart-product__info"
      >
        <img />
        <div
          className="cart-product__info__details"
        >
          <h2
            className="type--a7"
          >
            Product 1
          </h2>
          <p
            className="type--a8"
            style={
              Object {
                "marginTop": 5,
              }
            }
          >
            $10.00
          </p>
          <button
            className="cart-product__info__details__remove-btn type--a8"
            onClick={[Function]}
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        className="cart-product__quantity"
      >
        <button
          className="button type--a8  button--secondary button--dec"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Remove one item from the cart
          </span>
        </button>
        <span
          className="cart-product__quantity__value type--a8"
        >
          2
        </span>
        <button
          className="button type--a8  button--secondary-alt button--inc"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Add one item to cart
          </span>
        </button>
      </div>
    </li>
    <li
      className="cart-product"
    >
      <div
        className="cart-product__info"
      >
        <img />
        <div
          className="cart-product__info__details"
        >
          <h2
            className="type--a7"
          >
            Product 2
          </h2>
          <p
            className="type--a8"
            style={
              Object {
                "marginTop": 5,
              }
            }
          >
            $10.00
          </p>
          <button
            className="cart-product__info__details__remove-btn type--a8"
            onClick={[Function]}
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        className="cart-product__quantity"
      >
        <button
          className="button type--a8  button--secondary button--dec"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Remove one item from the cart
          </span>
        </button>
        <span
          className="cart-product__quantity__value type--a8"
        >
          2
        </span>
        <button
          className="button type--a8  button--secondary-alt button--inc"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Add one item to cart
          </span>
        </button>
      </div>
    </li>
    <li
      className="cart-product"
    >
      <div
        className="cart-product__info"
      >
        <img />
        <div
          className="cart-product__info__details"
        >
          <h2
            className="type--a7"
          >
            Product 3
          </h2>
          <p
            className="type--a8"
            style={
              Object {
                "marginTop": 5,
              }
            }
          >
            $10.00
          </p>
          <button
            className="cart-product__info__details__remove-btn type--a8"
            onClick={[Function]}
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        className="cart-product__quantity"
      >
        <button
          className="button type--a8  button--secondary button--dec"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Remove one item from the cart
          </span>
        </button>
        <span
          className="cart-product__quantity__value type--a8"
        >
          2
        </span>
        <button
          className="button type--a8  button--secondary-alt button--inc"
          disabled=""
          onClick={[Function]}
          type="button"
        >
          <span
            className="is-visually-hidden"
          >
            Add one item to cart
          </span>
        </button>
      </div>
    </li>
  </ul>
  <table
    className="border-section cart__breakdown type--a8"
    style={
      Object {
        "borderTopWidth": 1,
        "marginBottom": 80,
      }
    }
  >
    <tbody>
      <tr>
        <td
          className="cart__breakdown__label"
        >
          Subtotal
        </td>
        <td
          className="cart__breakdown__value"
        >
          $76.00
        </td>
      </tr>
      <tr>
        <td
          className="cart__breakdown__label"
        >
          Taxes
        </td>
        <td
          className="cart__breakdown__value"
        >
          $3.80
        </td>
      </tr>
      <tr
        className="border-section "
        style={
          Object {
            "borderTopWidth": 1,
          }
        }
      >
        <td
          className="cart__breakdown__label"
        >
          Total
        </td>
        <td
          className="cart__breakdown__value"
        >
          $79.80
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="button type--a8  button--primary cart__checkout-btn"
    disabled=""
    onClick={[Function]}
    type="button"
  >
    Checkout
  </button>
</div>
`;
